### PR TITLE
#174 Line mover for rule sections

### PR DESCRIPTION
**Pull request by ArtemDavletov**
Add correct Up/Down StatementMover for rule sections

Fixes: #174

### CODE REVIEW

File: src/test/kotlin/com/jetbrains/snakecharm/SmkStatementMoverTest.kt [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+    fun testInnerIf() {
+        doTest()
+    }
+
+    fun testNestedIfUp() {
+        doTest()
+    }
+
+    fun testCommentOut() {  //PY-5527
```
Comment (iromeo) [inline src/test/kotlin/com/jetbrains/snakecharm/SmkStatementMoverTest.kt:41]: What do all these tests? Do we really need to in-depth test all python line mover use cases here?

File: src/test/kotlin/com/jetbrains/snakecharm/SmkStatementMoverTest.kt [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+import com.intellij.openapi.actionSystem.IdeActions
+import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.jetbrains.python.psi.LanguageLevel
+import features.glue.SnakemakeWorld.myFixture
+
+class SmkStatementMoverTest: SnakemakeTestCase() {
+    private fun doTest() {
+        val testName: String = getTestName(true)
+
+        myFixture?.configureByFile("mover/$testName.smk")
```
Comment (iromeo) [inline src/test/kotlin/com/jetbrains/snakecharm/SmkStatementMoverTest.kt:12]: Test data is missing in the project.

Comment (iromeo) [inline src/test/kotlin/com/jetbrains/snakecharm/SmkStatementMoverTest.kt:12]: Let's rewrite this test case using cucumber. See `line_mover.feature` example

File: src/test/kotlin/com/jetbrains/snakecharm/SmkStatementMoverTest.kt (hunk 1 above)
Comment (ArtemDavletov) [reply src/test/kotlin/com/jetbrains/snakecharm/SmkStatementMoverTest.kt]: I removed this tests and added cucumber tests

File: src/test/kotlin/com/jetbrains/snakecharm/SmkStatementMoverTest.kt (hunk 2 above)
Comment (ArtemDavletov) [reply src/test/kotlin/com/jetbrains/snakecharm/SmkStatementMoverTest.kt]: Done

Comment (ArtemDavletov) [reply src/test/kotlin/com/jetbrains/snakecharm/SmkStatementMoverTest.kt]: Done

File: src/test/resources/features/actions/line_mover.feature [hunk 3]
```diff
@@ -1,20 +1,97 @@
...
-    When I put the caret at input
+    When I put the caret at <rule_like1> NAME1:
     And I invoke MoveStatementDown action
     Then editor content will be
     """
-    rule foo:
-        output:
-            "o"
-        input: "i"
-    """
```
Comment (iromeo) [inline src/test/resources/features/actions/line_mover.feature]: This test is an example of `input` section move, please support such case too.

**Comment by iromeo** [comment]
This issue is not only about "rules/checkpoints" blocks move up/down, but also rules sections, e.g. move "input" section block (with all its args up/down). Please update impl to cover such cases.

P.S. Also ensure that no exceptions occur in such case, e.g. at the moment:
For
```
# rule
rule rule174:
    input: "dd"

# fff
rule rule174:
    input: "dd"  # <-- caret here, press "Code | Move Statement Up"
    output:
        "boo"
```
I get:
```
Cannot perform move

com.intellij.diagnostic.PluginException: Wrong move ranges requested by com.jetbrains.python.codeInsight.editorActions.moveUpDown.PyStatementMover@1a411e87 58:126 vs 44:152 [Plugin: PythonCore]
	at com.intellij.ide.plugins.PluginManagerCore.createPluginException(PluginManagerCore.java:290)
	at com.intellij.diagnostic.PluginProblemReporterImpl.createPluginExceptionByClass(PluginProblemReporterImpl.java:12)
	at com.intellij.diagnostic.PluginException.createByClass(PluginException.java:58)
	at com.intellij.codeInsight.editorActions.moveUpDown.MoverWrapper.move(MoverWrapper.java:92)
	at com.intellij.codeInsight.editorActions.moveUpDown.BaseMoveHandler.executeWriteAction(BaseMoveHandler.java:57)
	at com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler$1.run(EditorWriteActionHandler.java:51)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:980)
	at com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler.doExecute(EditorWriteActionHandler.java:65)
	at com.intellij.openapi.editor.actionSystem.DynamicEditorActionHandler.doExecute(DynamicEditorActionHandler.java:52)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.lambda$execute$2(EditorActionHandler.java:191)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.doIfEnabled(EditorActionHandler.java:88)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.lambda$execute$3(EditorActionHandler.java:190)
	at com.intellij.openapi.editor.impl.CaretModelImpl.lambda$runForEachCaret$3(CaretModelImpl.java:299)
	at com.intellij.openapi.editor.impl.CaretModelImpl.doWithCaretMerging(CaretModelImpl.java:408)
	at com.intellij.openapi.editor.impl.CaretModelImpl.runForEachCaret(CaretModelImpl.java:308)
	at com.intellij.openapi.editor.impl.CaretModelImpl.runForEachCaret(CaretModelImpl.java:283)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.execute(EditorActionHandler.java:188)
	at com.intellij.openapi.editor.actionSystem.EditorAction.lambda$actionPerformed$0(EditorAction.java:89)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:220)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:177)
	at com.intellij.openapi.editor.actionSystem.EditorAction.actionPerformed(EditorAction.java:98)
	at com.intellij.openapi.editor.actionSystem.EditorAction.actionPerformed(EditorAction.java:73)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAware(ActionUtil.java:282)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher$1.performAction(IdeKeyEventDispatcher.java:610)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.lambda$processAction$3(IdeKeyEventDispatcher.java:670)
	at com.intellij.openapi.application.TransactionGuardImpl.performUserActivity(TransactionGuardImpl.java:94)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.java:669)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.java:620)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processActionOrWaitSecondStroke(IdeKeyEventDispatcher.java:517)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.inInitState(IdeKeyEventDispatcher.java:471)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.dispatchKeyEvent(IdeKeyEventDispatcher.java:220)
	at com.intellij.ide.IdeEventQueue.dispatchKeyEvent(IdeEventQueue.java:888)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:833)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:450)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:744)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:449)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:802)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:503)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

File: src/test/resources/features/actions/line_mover.feature (hunk 3 above)
Comment (ArtemDavletov) [reply src/test/resources/features/actions/line_mover.feature]: I added new tests.

File: src/test/resources/features/actions/line_mover.feature [hunk 4]
```diff
@@ -1,44 +1,59 @@
 Feature: Line mover
-  Scenario Outline: Move rules up
+  Scenario Outline: Permutation up/down between rules sections
```
Comment (iromeo) [inline src/test/resources/features/actions/line_mover.feature]: A permutation is a math term, means `a way, especially one of several possible variations, in which a set or number of things can be ordered or arranged.`. Please always check your translation in English-English dictionary. So here it sounds confusing. Please use here a more accurate translation, e.g. move up/down sections, swap rule like sections, etc

**Review by iromeo — CHANGES_REQUESTED** [review]
Tests don't cover all basic use cases. Please add more tests. E.g. not clear:
1) is it ok to move 'input' section out of the rule (as for wildcards example). 
2) could we move sections from one rule to the other rule, somewhat similar to move statements from one if branch to another.
3) Does mover works properly for 'run' section  and for python statements inside run section
4) Does run mover works for python statements inside snakemake files or it is implemented only for snakemake specific language constructions (e.g. rule etc sections)

File: src/test/resources/features/actions/line_mover.feature (hunk 4 above)
Comment (iromeo) [reply src/test/resources/features/actions/line_mover.feature]: P.S: I renamed tests

File: src/test/resources/features/actions/line_mover.feature [hunk 5]
```diff
@@ -1,44 +1,59 @@
...
+      | subworkflow | rule        | workdir        | output         | "file.txt"  | NAME1 | NAME2   | "file.txt" |
+      | rule        | checkpoint  | input          | output         | "file.txt"  | NAME1 | NAME2   | "file.txt" |
+      | checkpoint  | checkpoint  | input          | output         | "file.txt"  | NAME1 | NAME2   | "file.txt" |
+      | subworkflow | checkpoint  | workdir        | output         | "file.txt"  | NAME1 | NAME2   | "file.txt" |
+      | rule        | subworkflow | input          | workdir        | "file.txt"  | NAME1 | NAME2   | "file.txt" |
+      | checkpoint  | subworkflow | input          | workdir        | "file.txt"  | NAME1 | NAME2   | "file.txt" |
+      | subworkflow | subworkflow | workdir        | workdir        | "file.txt"  | NAME1 | NAME2   | "file.txt" |
+      | rule        | def         | input          | n              | "file.txt"  | NAME1 | NAME2() | int        |
+      | checkpoint  | def         | input          | n              | "file.txt"  | NAME1 | NAME2() | int        |
+      | subworkflow | def         | workdir        | n              | "file.txt"  | NAME1 | NAME2() | int        |
```
Comment (iromeo) [inline src/test/resources/features/actions/line_mover.feature:45]: Test data doesn't match name of the test, test ` Permutation up/down between rules sections` is supposed to be about rule sections swap, but `def` isn't `rule like` element and isn't rule section. Please split test in two or use better names for the test and test args to make test data matching test name.

**Review by iromeo — CHANGES_REQUESTED** [review]

File: src/test/resources/features/actions/line_mover.feature [hunk 6]
```diff
@@ -1,44 +1,59 @@
...
-      | subworkflow | rule        | workdir        | output         | "file.txt"      |
-      | rule        | checkpoint  | input          | output         | "file.txt"      |
-      | checkpoint  | checkpoint  | input          | output         | "file.txt"      |
-      | subworkflow | checkpoint  | workdir        | output         | "file.txt"      |
-      | rule        | subworkflow | input          | workdir        | "file.txt"      |
-      | checkpoint  | subworkflow | input          | workdir        | "file.txt"      |
-      | subworkflow | subworkflow | workdir        | workdir        | "file.txt"      |
-
-  Scenario Outline: Permutation up between rule and comment sections
+      | rule_like1  | rule_like2  | rule_section1  | rule_section2  | content1    | name1 | name2   | content2   |
```
Comment (iromeo) [inline src/test/resources/features/actions/line_mover.feature:33]: `content1` value is always `file.txt` here, `name1` is always `NAME1`. It is a but in test or redundant variables?

Comment (iromeo) [reply src/test/resources/features/actions/line_mover.feature:33]: Also you could use values like `input: "foo"` or `input:\n    "foo"` to make section with content single/multi line

File: src/test/resources/features/actions/line_mover.feature [hunk 7]
```diff
@@ -62,36 +89,213 @@ Feature: Line mover
       | checkpoint  | input          | ""      | # comment |
```
Comment (iromeo) [inline src/test/resources/features/actions/line_mover.feature:89]: Again what is the purpose of variables `comment` and `content` here?

File: src/test/resources/features/actions/line_mover.feature [hunk 8]
```diff
@@ -62,36 +89,213 @@ Feature: Line mover
...
+      | rule        | input          | output        | "file.txt"      | "file2.txt"      |
+      | checkpoint  | input          | output        | "file.txt"      | "file2.txt"      |
+      | subworkflow | workdir        | configfile    | "/"             | "/"              |
+
+  Scenario Outline: Permutation up/down inside rule section with complex content
     Given a snakemake project
     Given I open a file "foo1.smk" with text
     """
     <rule_like1> NAME1:
         <rule_section1>:
```
Comment (iromeo) [inline src/test/resources/features/actions/line_mover.feature:135]: This rule could be merged with the previous one if you use `\n` in content, e.g. you could use values like input: "foo" or input:\n "foo" to make section with content single/multi line

**Review by iromeo — CHANGES_REQUESTED** [review]

File: src/test/resources/features/actions/line_mover.feature [hunk 9]
```diff
@@ -413,3 +464,292 @@ Feature: Line mover
...
+      | rule        | rule        | \n"file1.txt"  | "file2.txt"   |           |
+      | checkpoint  | rule        | \n"file1.txt"  | "file2.txt"   |           |
+      | rule        | checkpoint  | \n"file1.txt"  | "file2.txt"   |           |
+      | checkpoint  | checkpoint  | \n"file1.txt"  | "file2.txt"   |           |
+      | checkpoint  | rule        | \n"file1.txt"  | \n"file2.txt" |           |
+      | rule        | checkpoint  | \n"file1.txt"  | \n"file2.txt" |           |
+      | checkpoint  | checkpoint  | \n"file1.txt"  | \n"file2.txt" |           |
+      | rule        | rule        | \n"file1.txt"  | \n"file2.txt" |           |
+
+  Scenario Outline: Swap rule between rule sections
```
Comment (iromeo) [inline src/test/resources/features/actions/line_mover.feature:680]: Scenario outline with the same name already defined above

File: src/test/resources/features/actions/line_mover.feature [hunk 10]
```diff
@@ -1,54 +1,78 @@
...
+      | checkpoint  | rule        | input          | output         | "file1.txt"    | "file2.txt"   |
+      | subworkflow | rule        | workdir        | output         | "file1.txt"    | "file2.txt"   |
+      | rule        | checkpoint  | input          | output         | "file1.txt"    | "file2.txt"   |
+      | checkpoint  | checkpoint  | input          | output         | "file1.txt"    | "file2.txt"   |
+      | subworkflow | checkpoint  | workdir        | output         | "file1.txt"    | "file2.txt"   |
+      | rule        | subworkflow | input          | workdir        | "file1.txt"    | "file2.txt"   |
+      | checkpoint  | subworkflow | input          | workdir        | "file1.txt"    | "file2.txt"   |
+      | subworkflow | subworkflow | workdir        | workdir        | "file1.txt"    | "file2.txt"   |
+      | rule        | def         | input          | n              | "file1.txt"    | int           |
+      | checkpoint  | def         | input          | n              | "file1.txt"    | int           |
```
Comment (iromeo) [inline src/test/resources/features/actions/line_mover.feature:41]: `def` isn't `rule like` snakemake section

File: src/main/kotlin/com/jetbrains/snakecharm/actions/SmkStatementMover.kt [hunk 11]
```diff
@@ -142,6 +142,27 @@ open class SmkStatementMover: PyStatementMover() {
         val statementList = getStatementList(elementToMove) ?: return true
         val statements = statementList.statements
 
+        if(elementToMove is SmkRuleOrCheckpointArgsSection) {
```
[the remark this answers is not in the document]
Comment (ArtemDavletov) [reply src/main/kotlin/com/jetbrains/snakecharm/actions/SmkStatementMover.kt:145]: I fixed

File: src/test/resources/features/actions/line_mover.feature (hunk 10 above)
Comment (ArtemDavletov) [reply src/test/resources/features/actions/line_mover.feature]: I added different test for python functions

File: src/test/resources/features/actions/line_mover.feature (hunk 7 above)
Comment (ArtemDavletov) [reply src/test/resources/features/actions/line_mover.feature]: Done

File: src/test/resources/features/actions/line_mover.feature (hunk 6 above)
Comment (ArtemDavletov) [reply src/test/resources/features/actions/line_mover.feature]: 'I put the caret at (.+)'  doesn't work correctly with line with \n

File: src/test/resources/features/actions/line_mover.feature (hunk 9 above)
Comment (ArtemDavletov) [reply src/test/resources/features/actions/line_mover.feature]: I changed it

File: src/main/kotlin/com/jetbrains/snakecharm/actions/SmkStatementMover.kt [hunk 12]
```diff
@@ -0,0 +1,543 @@
...
+                val text1 = document.getText(TextRange.create(start, end)).trim { it <= ' ' }
+                val text2 = document.getText(TextRange.create(document.getLineStartOffset(nearLine), document.getLineEndOffset(nearLine))).trim { it <= ' ' }
+                if (!text1.startsWith(quotes.first) && !text1.endsWith(quotes.second) && !text2.startsWith(quotes.first) && !text2.endsWith(quotes.second)) return true
+            }
+        }
+        return false
+    }
+
+    private fun getDestinationScope(file: PsiFile, editor: Editor,
+                                         elementToMove: PsiElement, down: Boolean): LineRange? {
```
[the remark this answers is not in the document]
Comment (ArtemDavletov) [reply src/main/kotlin/com/jetbrains/snakecharm/actions/SmkStatementMover.kt]: Done

File: src/test/resources/features/actions/line_mover.feature (hunk 5 above)
Comment (ArtemDavletov) [reply src/test/resources/features/actions/line_mover.feature]: I added different test for python functions

File: src/main/kotlin/com/jetbrains/snakecharm/actions/SmkStatementMover.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (iromeo) [reply src/main/kotlin/com/jetbrains/snakecharm/actions/SmkStatementMover.kt]: Not done, please read codestyle code conventions and reformat according to them.

File: src/test/resources/features/actions/line_mover.feature [hunk 13]
```diff
@@ -752,4 +791,222 @@ Feature: Line mover
...
+                  for f in job.subworkflow_input
+                  if job.subworkflow_input[f] is self
+              ]
+    """
+    When I put the caret at def targets(self, dag):
+    And I invoke MoveStatementUp action
+    Then editor content will be
+    """
+    rule NAME:
+      run:
```
Comment (iromeo) [inline src/test/resources/features/actions/line_mover.feature:958]: Do you really need such long test data? Please use minimal test case here.

**Review by iromeo — CHANGES_REQUESTED** [review]

**Comment by iromeo** [comment]
Implementation contain lot's of copy-paste from `PyStatementMover` w/o any evaluation what code in `PyStatementMover` should become a public API. This code should be significantly refactored with `PyStatementMover`

**Pull request merged**